### PR TITLE
HIVE-29514: Optimize UDF Unhex and improve its test coverage

### PR DESCRIPTION
**Pull request by tanishq-chugh**
### What changes were proposed in this pull request?

Improve the UDF Unhex and its test coverage

### Why are the changes needed?

Better performance

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Manual testing + java test

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java [hunk 1]
```diff
@@ -42,32 +42,59 @@
...
+
+    int i = 0;
+    int resIdx = 0;
+
+    // If length is odd, the first character acts as the first byte avoiding adding "0" prefix
+    if (len % 2 != 0) {
+      int val = decodeHexChar(textBytes[i++]);
+      if (val == -1) {
+        return null;
+      }
```
Comment (soumyakanti3578) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java:70]: Please add a test for invalid character in an odd-length input.
Example: `G`.

File: ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java [hunk 2]
```diff
@@ -42,32 +42,59 @@
...
+    while (i < len) {
+      int high = decodeHexChar(textBytes[i++]);
+      int low  = decodeHexChar(textBytes[i++]);
+
+      if (high == -1 || low == -1) {
         return null;
       }
+
+      result[resIdx++] = (byte) ((high << 4) | low);
     }
```
Comment (soumyakanti3578) [inline ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java:81]: How about:
```
while (i < len) {
  int high, low;
  if ((high = decodeHexChar(textBytes[i++])) == -1 ||
      (low  = decodeHexChar(textBytes[i++])) == -1) {
    return null;
  }
  result[resIdx++] = (byte) ((high << 4) | low);
}
```
so that we don't compute `low` if `high == -1`. Consider adding tests for when `high == -1` and `low == -1`.

File: ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFUnhex.java [hunk 3]
```diff
@@ -40,8 +42,41 @@ public void testUnhexConversion(){
...
+  @Test
+  public void testUnhexNullEmptyCases() {
+    UDFUnhex udf = new UDFUnhex();
+
+    assertNull(udf.evaluate(null));
+
+    Text hexEmpty = new Text("");
+    byte[] expectedEmpty = new byte[0];
+    assertArrayEquals(expectedEmpty, udf.evaluate(hexEmpty));
   }
```
Comment (soumyakanti3578) [inline ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFUnhex.java:91]: It would be nice to have tests for 

- mixed case: `aABb9`
- boundary values
- lower case, as all tests are for upper case right now

Also, maybe not in this file, but are there any tests for `hex(unhex(...))` and `unhex(hex(..))`?

**Review by thomasrebele — APPROVED** [review]
I agree with @soumyakanti3578's comments. Otherwise LGTM.

File: ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java (hunk 1 above)
Comment (tanishq-chugh) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java:70]: Hi @soumyakanti3578 , Thanks for checking this!
Added the case in commit: [a530e08](https://github.com/apache/hive/pull/6471/commits/a530e083b55cf8ee5df198698d28a41b709e9329)

File: ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java (hunk 2 above)
Comment (tanishq-chugh) [reply ql/src/java/org/apache/hadoop/hive/ql/udf/UDFUnhex.java:81]: that would be better for optimisation, have updated the loop in commit: [318630a](https://github.com/apache/hive/pull/6471/commits/318630a163306c926b9cd6eb4dd5a43c8d6c08fb)

Have added the test cases for high/low in commit: [d93c476](https://github.com/apache/hive/pull/6471/commits/d93c476db0d893fd620b9d96e5e0076002df1f57)

File: ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFUnhex.java (hunk 3 above)
Comment (tanishq-chugh) [reply ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFUnhex.java:91]: Have added these test cases in commit: [254027f](https://github.com/apache/hive/pull/6471/commits/254027f42517e7e1fda76d355c1ef31aade3b4ae)

>Also, maybe not in this file, but are there any tests for hex(unhex(...)) and unhex(hex(..))?

I see the `UNHEX(HEX(...))` case being used in `ba_table_udfs.q` QTest, but i don't think `HEX(UNHEX(...))` is a valid scenario, what do you think? please feel free to correct me if i am wrong here.

Comment (soumyakanti3578) [reply ql/src/test/org/apache/hadoop/hive/ql/udf/TestUDFUnhex.java:91]: Yeah I don't think `HEX(UNHEX(...))` is really valid.

**Review by soumyakanti3578 — APPROVED** [review]
LGTM 👍🏼
